Remora.Commands/Trees/CommandTreeBuilder.cs

```diff
@@ -190,14 +277,114 @@ private IEnumerable<CommandNode> GetModuleCommands(Type moduleType, IParentNode
                 );
             }
 
+            method.GetAttributesAndConditions(out var attributes, out var conditions);
+
+            if (isInUnnamedGroup)
+            {
+                // If the group is being hoisted, take the attributes of the parent type(s).
+                ExtractExtraAttributes(moduleType, out var extraAttributes, out var extraConditions);
+
+                attributes = extraAttributes.Concat(attributes).ToArray();
+                conditions = extraConditions.Concat(conditions).ToArray();
+            }
+
             yield return new CommandNode
             (
                 parent,
                 commandAttribute.Name,
-                moduleType,
-                method,
-                commandAttribute.Aliases
+                CreateDelegate(method, method.GetParameters().Select(p => p.ParameterType).ToArray()),
+                CommandShape.FromMethod(method),
+                commandAttribute.Aliases,
+                attributes,
+                conditions
             );
         }
     }
+
+    private static Func<IServiceProvider, object?[], CancellationToken, ValueTask<IResult>> CreateDelegate(MethodInfo method, Type[] argumentTypes)
+    {
+        // Get the object from the container
+        var serviceProvider = Expression.Parameter(typeof(IServiceProvider), "serviceProvider");
+
+        var instance = Expression.Call(serviceProvider, GetServiceMethodInfo, Expression.Constant(method.DeclaringType));
+
+        var parameters = Expression.Parameter(typeof(object?[]), "parameters");
+
+        // Create the arguments
+        var arguments = new Expression[argumentTypes.Length];
+        for (var i = 0; i < argumentTypes.Length; i++)
+        {
+            var argumentType = argumentTypes[i];
+            var argument = Expression.ArrayIndex(parameters, Expression.Constant(i));
+            arguments[i] = Expression.Convert(argument, argumentType);
+        }
+
+        var castedInstance = Expression.Convert(instance, method.DeclaringType!);
+
+        var call = Expression.Call(castedInstance, method, arguments);
+
+        // Convert the result to a ValueTask<IResult>
+        call = (MethodCallExpression)CoerceToValueTask(call);
```

Comment:
Pass expressionType to CoerceToValueTask if possible; in this case I think it's `method.ReturnType`

Reply:
Also this cast is silly, make CoerceToValueTask return MethodCallExpression instead.

Reply:
Perhaps it is daft. Also what was the benefit of passing the type?

Reply:
> what was the benefit of passing the type?

Looking harder at it, you probably don't need to do that, actually